File: widgetssdk/src/main/java/com/glia/widgets/view/unifieduiconfig/component/base/Size.java
package com.glia.widgets.view.unifieduiconfig.component.base;

import androidx.annotation.Dimension;

import com.glia.widgets.view.unifieduiconfig.deserializer.DpDeserializer;
import com.glia.widgets.view.unifieduiconfig.deserializer.SpDeserializer;

/**
 * Represents density independent pixels
 */
public interface Size {

    /**
     * @return size in pixels
     */
    @Dimension
    float getSizePx();

    default int getIntSizePx() {
        return Math.round(getSizePx());
    }

    /**
     * Represents size in Sp (Scalable pixel)
     * Used by deserializer to differ sp from dp
     *
     * @see SpDeserializer
     */
    interface Sp extends Size {
    }

    /**
     * Represents size in Dp (Density independent pixel)
     * Used by deserializer to differ sp from dp
     *
     * @see DpDeserializer
     */
    interface Dp extends Size {
    }
}

File: widgetssdk/src/main/java/com/glia/widgets/view/unifiedui/config/base/Size.kt
package com.glia.widgets.view.unifiedui.config.base

import kotlin.math.roundToInt


/**
 * Represents size in Sp (Scalable pixel)
 * Used by deserializer to differ sp from dp
 *
 * @see [com.glia.widgets.view.unifiedui.parse.SpDeserializer]
 */
@JvmInline
internal value class SizeSpRemoteConfig(val value: Float) {
    val intValue: Int
        get() = value.roundToInt()
}

/**
 * Represents size in Dp (Density independent pixel)
 * Used by deserializer to differ sp from dp
 *
 * @see [com.glia.widgets.view.unifiedui.parse.DpDeserializer]
 */
@JvmInline
internal value class SizeDpRemoteConfig(val valuePx: Float) {
    val intValuePx: Int
        get() = valuePx.roundToInt()
}